src/LendController.sol

```diff
@@ -0,0 +1,156 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity 0.8.13;
+
+import {ERC721, ERC721TokenReceiver} from 'solmate/tokens/ERC721.sol';
+import {SafeTransferLib, ERC20} from "solmate/utils/SafeTransferLib.sol";
+
+import {INFTLoanFacilitator} from './interfaces/INFTLoanFacilitator.sol';
+
+struct LoanTerms {
+    bool allowBuyouts;
+    bool allowLoanAmountIncrease;
+    uint16 interestRate;
+    uint128 amount;
+    uint32 durationSeconds;
+    address loanAsset;
+}
+
+interface NFTLoanTermsSource {
+    function terms(uint256 tokenId, address nft, address loanAsset, bytes calldata data) external returns (LoanTerms memory);
+}
+
+contract LendController is ERC721TokenReceiver {
+    using SafeTransferLib for ERC20;
+    
+    struct LenderSpec {
+        address lender;
+        uint256 amount;
+        bool pull;
+    }
+
+    struct BorrowRequest {
+        LenderSpec[] lendersSpecs;
+        address termsSource;
+        LoanTerms terms;
```

Comment:
surprised the borrow request specifies the terms rather than it being derived from the terms source, but makes sense so the terms returned from the source can be validated. think these likely should be diff structs though

Reply:
yeah was concerned about a race condition where terms contract is changed while borrow request tx is in the mempool. Agree could be different structs.